src/terminal/parser/InputStateMachineEngine.cpp

```diff
@@ -454,6 +462,26 @@ bool InputStateMachineEngine::ActionCsiDispatch(const VTID id, const VTParameter
     case CsiActionCodes::FocusOut:
         success = _pDispatch->FocusChanged(false);
         break;
+    case CsiActionCodes::DA_DeviceAttributes:
+        // This assumes that InputStateMachineEngine is tightly coupled with VtInputThread and the rest of the ConPTY system.
+        // On startup, ConPTY will send a DA1 request to get more information about the hosting terminal.
+        // We catch it here and store the information for later retrieval.
+        if (_deviceAttributes.load(std::memory_order_relaxed) == 0)
+        {
+            til::enumset<DeviceAttribute, uint64_t> attributes;
+
+            // The first parameter denotes the conformance level, so we skip it.
+            parameters.subspan(1).for_each([&](auto p) {
+                attributes.set(static_cast<DeviceAttribute>(p));
+                return true;
+            });

```

Comment:
In practice this probably won't matter, but strictly speaking we should be checking that the first parameter is 60+, otherwise we can't assume that we're getting a list of features here. When the first parameter is less than 60, the parameters that follow take on a different meaning.

Reply:
Should I check for >=60 or >=61? Documentation and comments that I found here imply that all terminals that support DA1 are at least 61...

Reply:
Using >=61 is fine, but I don't think it would matter either way. In STD070 they document the range 61 to 69 as representing levels 1 to 9. But they also go on say that values 1 to 59 are device specific codes, and that the new encoding scheme applies to values 60 or greater. So it's a bit ambiguous, but it's highly unlikely we'll ever encounter a terminal with the value 60 either way.

And while we're getting into technicalities, I should mention that certain features are supposed to be implicit at higher levels. In particular level 4+ devices are supposed to support the rectangular editing operations, but won't advertise feature 28 explicitly. Unfortunately modern terminals tend to lie about their conformance level, so we can't trust that value, and we're better off assuming no support unless feature 28 is explicitly listed.